### DISCUSSION TITLE
DJL Monthly Update - February 2021

### DISCUSSION THREAD

**Discussion by zachgk** [deepjavalibrary/djl#737] [News]
## Community

**The DJL team uploaded a DJL 101 video series. It is about 1.5 hours total (8 videos of ~10 minutes each) and covers setup, ndarrays, training, and inference https://www.youtube.com/watch?v=J4wUnxOYN8M&list=PLC1JzXeHJitDmZBHupMGZE2zHzU8fEvfD&index=2&t=2s**

Patrick Arnold published an article titled "AWS Deep Java Library – Was ist das?" in AUSY (written in German) https://www.ausy-technologies.de/de/insight/aws-deep-java-library---was-ist-das-~i3706

@lanking520 published the blog post "Train “Undying” Flappy Bird Using Reinforcement Learning on Java" https://towardsdatascience.com/train-undying-flappy-bird-using-reinforcement-learning-on-java-98ff68eb28bf

NVIDIA published a blog post on using DJL on GPU-powered Spark for running inference on deep learning models https://developer.nvidia.com/blog/accelerating-deep-learing-with-apache-spark-and-gpus-on-aws/

### Events

@frankliu and @skirdey presented on DJL to the *Chicago JUG* on 2/18 https://www.youtube.com/watch?v=gBy99IhjSSA.

@roywei and @skirdey presented in the St. Louis *Apache Kafka* Meetup on using DJL to deploy models from streaming applications on March 3rd https://www.meetup.com/Saint-Louis-Kafka-meetup-group/events/276466753/?isFirstPublish=true

@goswamig and @skirdey will present on deploying deep learning models using DJL from Scala-based big data pipelines in the Bay Area Big Data and Cloud Meetup group on March 30 https://www.meetup.com/datariders/events/276592801/

## Feature Enhancements

* Paddle 2.0.0 published
* Model serving auto scale feature
* **Make default engine deterministic when multiple engines are available**
* Implemented the norm operator
* Add DJL serving worker idle time, (https://github.com/awslabs/djl/pull/613)
* Improve the PyTorch index performance (https://github.com/awslabs/djl/pull/638)
* Improvements for MultiThreadedBenchmark, (https://github.com/awslabs/djl/pull/617)
* Add support for java-9 automatic module name, (https://github.com/awslabs/djl/pull/627)
* Add PaddlePaddle OCR notebook (https://github.com/awslabs/djl/pull/640)
* **List input and output classes when loading model zoo model (https://github.com/awslabs/djl/pull/624)**
* Make XavierInitializer by default for all weights and improve setInitializer (https://github.com/awslabs/djl/pull/664)
* Updated SageMaker notebook docker image to support execute system command
* Adding Central add metadata info, (https://github.com/awslabs/djl/pull/674)
* Make embedding operator generic across engines (https://github.com/awslabs/djl/pull/688)
* Added Symbolic Model loading and saving mechanism (https://github.com/awslabs/djl/pull/687) (https://github.com/awslabs/djl/pull/669)
* **Adds a helper ObjectDetectionDataset (https://github.com/awslabs/djl/pull/683)**

## Documentation & Examples

* [PaddlePaddle] OCR example (https://github.com/awslabs/djl/pull/568)
* Updated tutorial and miscellaneous documentation (https://github.com/awslabs/djl/pull/607)
* Edited community page #576 (https://github.com/awslabs/djl/pull/576/files)
* Added demo project to show case who to build java-9 modular application with DJL
* Added SageMaker Studio Notebook setup for DJL
* Document DJL forums (https://github.com/awslabs/djl/pull/646)
* Add guide to engines (https://github.com/awslabs/djl/pull/660)
* Document applications, link from Javadoc to D2L, fix tutorial (https://github.com/awslabs/djl/pull/673)

### D2L Book

* [D2L] Complete Section 2.4: Calculus (https://github.com/aws-samples/d2l-java/commit/c501d0700ee4432aa54bc94f51edfd1b1be43ac2)
* [D2L] Complete Section 2.5: Automatic Differentiation (https://github.com/aws-samples/d2l-java/pull/80)
* [D2L] Completed section 2.6: Probability  (https://github.com/aws-samples/d2l-java/pull/82)
* [D2L] Complete Section 2.7: Documentation (https://github.com/aws-samples/d2l-java/pull/81)
* [D2L] Completed section 8.1: Sequence Models  (https://github.com/aws-samples/d2l-java/pull/83)
* [D2L] Completed section 8.2: Sequence Models (https://github.com/aws-samples/d2l-java/pull/84)
* [D2L] Completed section 8.3: Language Models and the Dataset (https://github.com/aws-samples/d2l-java/pull/85)
* [D2L] Completed section 8.4: Recurrent Neural Networks (https://github.com/aws-samples/d2l-java/pull/86)

## Bug Fixes

* Rename CheckpointsTrainingListener to SaveModelTrainingListener (https://github.com/awslabs/djl/pull/573)
* Fixed model zoo supported engine bug for DLR, TFLite and OnnxRuntime model zoo
* Upgraded maven wrapper script for Windows to 0.5.6
* Fixed gradle wrapper TLS issue for Windows
* Upgrade TFLite to 2.4.1
* Fixed Model View bug to prevent crashing on a null model value
* Refactor the initialize workflow for easier customization (https://github.com/awslabs/djl/pull/675)
* Fixed broken jupyter notebook
* Refactor BERT block to remove unnecessary overloading forward (https://github.com/awslabs/djl/pull/696)
* Refactored PyTorch pipeline to remove extra upload stage
* Fixed model zoo supported engine bug (https://github.com/awslabs/djl/pull/616)
* Fixed broken jupyter notebooks
* Fixed PyTorch training performance issue (https://github.com/awslabs/djl/issues/621)
* Investigated and resolved IntelliJ plugin classloader question (https://github.com/awslabs/djl/issues/622)
* Investigated and resolved java.nio.ReadOnlyBufferException (https://github.com/awslabs/djl/issues/570)
* Fixed softmax from_logit flag and RNN tests (https://github.com/awslabs/djl/pull/639)
* Fixed Jacoco code coverage report to reflect real coverage (exclude example and integration code)
* Fixed PaddlePaddle engine co-exist with MXNet issue
* Fix formatCpp linter (https://github.com/awslabs/djl/pull/682)
* Enabled Codecov report for DJL (https://github.com/awslabs/djl/pull/679) 
* Updates BERT example including testing (https://github.com/awslabs/djl/pull/681)
* Removes old DataManager class (https://github.com/awslabs/djl/pull/691)
